test/indexer/test_tdr.py

```diff
@@ -205,6 +341,35 @@ def _run_sql(self, query: str) -> BigQueryRows:
     def _full_table_name(self, source: TDRSourceName, table_name: str) -> str:
         return source.bq_name + '.' + table_name
 
+    def _find_upstream_bundles(self,
+                               source: TDRSourceRef,
+                               outputs: Entities
+                               ) -> Set[SourcedBundleFQID]:
+        # TinyQuery/legacy SQL has no support for BQ Arrays, so it's impossible
+        # to test the query in the overridden method.
+        all_links = self._run_sql(f'''
+            SELECT links_id, version, content
+            FROM {self._full_table_name(source.name, 'links')}
+        ''')
+        bundles = {}
+        for row in all_links:
+            outputs_found = outputs & {
+                EntityReference(entity_id=output['output_id'],
+                                entity_type=output['output_type'])
+                for link in json.loads(row['content'])['links']
+                if link['link_type'] == 'process_link'
+                for output in link['outputs']
+            }
+            if outputs_found:
+                bundles[SourcedBundleFQID(uuid=row['links_id'],
+                                          version=self.format_version(row['version']),
+                                          source=source)] = outputs_found
+
+        missing = outputs - set.union(set(), *bundles.values())
```

Comment:
why not just

```suggestion
        missing = outputs - set(bundles.values())
```

or even

```suggestion
        missing = outputs.difference(bundles.values())
```

Reply:
This snippet is baffling to me. I think the `set()` is a typo. However, the call to `union` is necessary since the `values` of `bundles` are themselves sets. I think the variable ought to be named `outputs_by_bundle` and ought to have a type annotation.

Reply:
Ahh, ok, so

```suggestion
        missing = outputs - set.union(*bundles.values())
```

Reply:
or 

```suggestion
        missing = outputs - set.union(set(), *outputs_by_bundle.values())
```

Reply:
I was mistaken about `set()` being a typo. The original code actually was intended and necessary. Without the first empty set, the code raises an exception if `outputs_by_bundle` is empty, because `set.union` expects at least one argument.